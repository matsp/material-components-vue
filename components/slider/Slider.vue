<template>
  <div
    :aria-disabled="disabled"
    :aria-label="ariaLabel"
    :class="classes"
    class="mdc-slider"
    role="slider"
    tabindex="0"
    @MDCSlider:change="onChange"
  >
    <div class="mdc-slider__track-container">
      <div class="mdc-slider__track" />
      <div
        v-if="displayMarkers"
        class="mdc-slider__track-marker-container"
      />
    </div>
    <div
      v-if="discrete"
      class="mdc-slider__thumb-container"
    >
      <div class="mdc-slider__pin">
        <span class="mdc-slider__pin-value-marker" />
      </div>
      <svg
        class="mdc-slider__thumb"
        width="21"
        height="21"
      >
        <circle
          cx="10.5"
          cy="10.5"
          r="7.875"
        />
      </svg>
      <div class="mdc-slider__focus-ring" />
    </div>
    <div
      v-else
      class="mdc-slider__thumb-container"
    >
      <svg
        class="mdc-slider__thumb"
        width="21"
        height="21"
      >
        <circle
          cx="10.5"
          cy="10.5"
          r="7.875"
        />
      </svg>
      <div class="mdc-slider__focus-ring" />
    </div>
  </div>
</template>

<script>
import { MDCSlider } from '@material/slider'

import { baseComponentMixin, themeClassMixin } from '../base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  model: {
    prop: 'value',
    event: 'change'
  },
  props: {
    ariaLabel: {
      type: String,
      default: ''
    },
    displayMarkers: {
      type: Boolean,
      default: false
    },
    discrete: {
      type: Boolean,
      default: false
    },
    value: {
      type: Number,
      default: 0
    },
    min: {
      type: Number,
      default: 0
    },
    max: {
      type: Number,
      default: 100
    },
    step: {
      type: Number,
      validator: value => value > 0,
      default: 1
    },
    disabled: {
      type: Boolean,
      default: false
    }
  },
  data () {
    return {
      mdcSlider: undefined
    }
  },
  computed: {
    classes () {
      return {
        'mdc-slider--discrete': this.discrete,
        'mdc-slider--display-markers': this.displayMarkers
      }
    }
  },
  watch: {
    value () {
      this.mdcSlider.value = this.value
    },
    min () {
      this.mdcSlider.min = this.min
    },
    max () {
      this.mdcSlider.max = this.max
    },
    step () {
      this.mdcSlider.step = this.step
    },
    disabled () {
      this.mdcSlider.disabled = this.disabled
    }
  },
  mounted () {
    this.mdcSlider = MDCSlider.attachTo(this.$el)
    this.mdcSlider.value = this.value
    this.mdcSlider.min = this.min
    this.mdcSlider.max = this.max
    this.mdcSlider.step = this.step
    this.mdcSlider.disabled = this.disabled
  },
  beforeDestroy () {
    this.mdcSlider.destroy()
  },
  methods: {
    onChange (event) {
      this.$emit('change', this.mdcSlider.value)
    }
  }
}
</script>
