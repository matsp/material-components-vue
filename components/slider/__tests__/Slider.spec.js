import { mount } from '@vue/test-utils'
import Slider from '../Slider.vue'

describe('Slider', () => {
  it('should mount', () => {
    const wrapper = mount(Slider)
    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper.vm.$data.mdcSlider).toBeDefined()
  })

  it('should render with no prop', () => {
    const wrapper = mount(Slider)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-slider')
  })

  it('should render as discrete', () => {
    const wrapper = mount(Slider, {
      propsData: {
        discrete: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-slider--discrete')
  })

  it('should render and display markers', () => {
    const wrapper = mount(Slider, {
      propsData: {
        discrete: true,
        displayMarkers: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-slider--display-markers')
  })

  it('should render as disabled', () => {
    const wrapper = mount(Slider, {
      propsData: {
        disabled: true
      }
    })
    expect(wrapper).toMatchSnapshot()
  })
})
