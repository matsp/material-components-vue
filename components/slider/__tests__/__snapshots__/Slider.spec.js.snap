// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider should render and display markers 1`] = `
<div role="slider" tabindex="0" class="mdc-slider mdc-slider--discrete mdc-slider--display-markers" aria-valuenow="0" aria-valuemin="0" aria-valuemax="100">
  <div class="mdc-slider__track-container">
    <div class="mdc-slider__track"></div>
    <div class="mdc-slider__track-marker-container"></div>
  </div>
  <div class="mdc-slider__thumb-container">
    <div class="mdc-slider__pin"><span class="mdc-slider__pin-value-marker"></span></div> <svg width="21" height="21" class="mdc-slider__thumb">
      <circle cx="10.5" cy="10.5" r="7.875"></circle>
    </svg>
    <div class="mdc-slider__focus-ring"></div>
  </div>
</div>
`;

exports[`Slider should render as disabled 1`] = `
<div aria-disabled="true" role="slider" class="mdc-slider mdc-slider--disabled" aria-valuenow="0" aria-valuemin="0" aria-valuemax="100">
  <div class="mdc-slider__track-container">
    <div class="mdc-slider__track"></div>
    <!---->
  </div>
  <div class="mdc-slider__thumb-container"><svg width="21" height="21" class="mdc-slider__thumb">
      <circle cx="10.5" cy="10.5" r="7.875"></circle>
    </svg>
    <div class="mdc-slider__focus-ring"></div>
  </div>
</div>
`;

exports[`Slider should render as discrete 1`] = `
<div role="slider" tabindex="0" class="mdc-slider mdc-slider--discrete" aria-valuenow="0" aria-valuemin="0" aria-valuemax="100">
  <div class="mdc-slider__track-container">
    <div class="mdc-slider__track"></div>
    <!---->
  </div>
  <div class="mdc-slider__thumb-container">
    <div class="mdc-slider__pin"><span class="mdc-slider__pin-value-marker"></span></div> <svg width="21" height="21" class="mdc-slider__thumb">
      <circle cx="10.5" cy="10.5" r="7.875"></circle>
    </svg>
    <div class="mdc-slider__focus-ring"></div>
  </div>
</div>
`;

exports[`Slider should render with no prop 1`] = `
<div role="slider" tabindex="0" class="mdc-slider" aria-valuenow="0" aria-valuemin="0" aria-valuemax="100">
  <div class="mdc-slider__track-container">
    <div class="mdc-slider__track"></div>
    <!---->
  </div>
  <div class="mdc-slider__thumb-container"><svg width="21" height="21" class="mdc-slider__thumb">
      <circle cx="10.5" cy="10.5" r="7.875"></circle>
    </svg>
    <div class="mdc-slider__focus-ring"></div>
  </div>
</div>
`;
